`crucible-llvm`: Support vectorized `llvm.u{min,max}` and `llvm.s{min,max}` intrinsics

`crux-llvm`'s [`T1177.c`](https://github.com/GaloisInc/crucible/blob/b9bd2bac457568875423c017dd6113a6162e6f06/crux-llvm/test-data/golden/T1177.c) test case passes with Clang 14, but not with Clang 16 or later. Some investigation reveals that this is because Clang 16+ is more likely to optimize code to uses of vectorized [`llvm.umin`](https://llvm.org/docs/LangRef.html#llvm-umin-intrinsic) intrinsics (as well as its cousins, vectorized [`llvm.umax`](https://llvm.org/docs/LangRef.html#llvm-vector-reduce-umax-intrinsic), [`llvm.smin`](https://llvm.org/docs/LangRef.html#llvm-vector-reduce-smin-intrinsic), and [`llvm.smax`](https://llvm.org/docs/LangRef.html#llvm-vector-reduce-smax-intrinsic)). Here is a minimized version of `T1177.c` that demonstrates the issue with vectorized `llvm.umin` in particular:

```c
// test.c
#include <crucible.h>
#include <stddef.h>
#include <stdint.h>

uint32_t __attribute__((noinline)) reduce_umin(const uint8_t* in, size_t len) {
  uint32_t x = UINT32_MAX;

  for (size_t i = 0; i < len; i++) {
    uint32_t y = in[i];
    x = (x < y) ? x : y;
  }

  return x;
}

int main(void) {
  const uint8_t unsigned_in[8] = {0, 1, 2, 3, 4, 5, 6, 7};
  check(reduce_umin(unsigned_in, 8) == 0);
  return 0;
}
```

`crux-llvm` can simulate this with Clang 14:

```
$ CLANG=clang-14 cabal -v0 run exe:crux-llvm -- -O2 test.c
[Crux] Using pointer width: 64 for file crux-build/crux~test.bc
[Crux] Simulating function main
[Crux] Attempting to prove verification conditions.
[Crux] Goal status:
[Crux]   Total: 1
[Crux]   Proved: 1
[Crux]   Disproved: 0
[Crux]   Incomplete: 0
[Crux]   Unknown: 0
[Crux] Overall status: Valid.
```

With Clang 16 or later, however, it fails:

```
$ CLANG=clang-18 cabal -v0 run exe:crux-llvm -- -O2 test.c
[Crux] Using pointer width: 64 for file crux-build/crux~test.bc
[Crux] Simulating function main
[Crux] Attempting to prove verification conditions.
[Crux] *** debug executable: results/test/debug-11
[Crux] *** break on line: 11
[Crux] Found counterexample for verification goal
[Crux]   test.c:11:9: error: in reduce_umin
[Crux]   Failed to load function handle
[Crux]   Details:
[Crux]     No implementation or override found for pointer
[Crux]     The given pointer could not be resolved to a callable function
[Crux]     No implementation or override found for pointer
[Crux]     Attempting to load callable function with type: <4 x i32>(<4 x i32>, <4 x i32>)
[Crux]       Via pointer: Global symbol "llvm.umin.v4i32" (7, 0x0:[64])
[Crux]     In memory state:
[Crux]       Stack frame reduce_umin
[Crux]         No writes or allocations
[Crux]       Stack frame main
[Crux]         Allocations:
[Crux]           StackAlloc 12 0x8:[64] Mutable 8-byte-aligned internal
[Crux]         Writes:
[Crux]           Indexed chunk:
[Crux]             12 |->   *(12, 0x0:[64]) := 0x706050403020100:[64]
[Crux]       Base memory
[Crux]         Allocations:
[Crux]           GlobalAlloc 11 0x7:[64] Immutable 1-byte-aligned [global variable  ] .str
[Crux]           GlobalAlloc 10 0x0:[64] Immutable 1-byte-aligned [defined function ] main
[Crux]           GlobalAlloc 9 0x0:[64] Immutable 1-byte-aligned [defined function ] reduce_umin
[Crux]           GlobalAlloc 8 0x0:[64] Immutable 1-byte-aligned [external function] llvm.vector.reduce.umin.v4i32
[Crux]           GlobalAlloc 7 0x0:[64] Immutable 1-byte-aligned [external function] llvm.umin.v4i32
[Crux]           GlobalAlloc 6 0x0:[64] Immutable 1-byte-aligned [external function] llvm.umin.i32
[Crux]           GlobalAlloc 5 0x0:[64] Immutable 1-byte-aligned [external function] llvm.dbg.value
[Crux]           GlobalAlloc 4 0x0:[64] Immutable 1-byte-aligned [external function] llvm.dbg.assign
[Crux]           GlobalAlloc 3 0x0:[64] Immutable 1-byte-aligned [external function] crucible_assert
[Crux]           GlobalAlloc 2 0x0:[64] Immutable 1-byte-aligned [external function] llvm.lifetime.end.p0
[Crux]           GlobalAlloc 1 0x0:[64] Immutable 1-byte-aligned [external function] llvm.lifetime.start.p0
[Crux]         Writes:
[Crux]           Indexed chunk:
[Crux]             11 |->   *(11, 0x0:[64]) := "test.c\NUL"
[Crux]     in context:
[Crux]       reduce_umin
[Crux]       main
[Crux] Goal status:
[Crux]   Total: 1
[Crux]   Proved: 0
[Crux]   Disproved: 1
[Crux]   Incomplete: 0
[Crux]   Unknown: 0
[Crux] Overall status: Invalid.
```

The problematic intrinsic has this type signature:

```llvm
declare <4 x i32> @llvm.umin.v4i32(<4 x i32>, <4 x i32>)
```

We will need to add a `crucible-llvm` override for this. ([There are already overrides for the unvectorized versions.](https://github.com/GaloisInc/crucible/blob/b9bd2bac457568875423c017dd6113a6162e6f06/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/LLVM.hs#L241-L252))

There are of course many more vectorized intrinsics that `crucible-llvm` could conceivably support, but for now, I propose that we limit ourselves to what we need to make the `crux-llvm` test suite pass.

--- a/crucible-llvm/CHANGELOG.md
+++ b/crucible-llvm/CHANGELOG.md
@@ -13,6 +13,8 @@
 * Fix a bug that would cause the simuator to compute incorrect results for the
   `llvm.is.fpclass` intrinsic. (Among other things, this is used to power the
   `isnan` function in Clang 17 or later.)
+* Support vectorized versions of the `llvm.u{min,max}` and `llvm.s{min,max}`
+  intrinsics.
 
 # 0.7.1 -- 2025-03-21
 

--- a/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/LLVM.hs
+++ b/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/LLVM.hs
@@ -307,6 +307,23 @@ poly1_vec_llvm_overrides =
     , Poly1VecLLVMOverride $ \vecSz intSz ->
         SomeLLVMOverride (llvmVectorReduceUmin vecSz intSz)
     )
+
+  , ("llvm.smax"
+    , Poly1VecLLVMOverride $ \vecSz intSz ->
+        SomeLLVMOverride (llvmSmaxVector vecSz intSz)
+    )
+  , ("llvm.smin"
+    , Poly1VecLLVMOverride $ \vecSz intSz ->
+        SomeLLVMOverride (llvmSminVector vecSz intSz)
+    )
+  , ("llvm.umax"
+    , Poly1VecLLVMOverride $ \vecSz intSz ->
+        SomeLLVMOverride (llvmUmaxVector vecSz intSz)
+    )
+  , ("llvm.umin"
+    , Poly1VecLLVMOverride $ \vecSz intSz ->
+        SomeLLVMOverride (llvmUminVector vecSz intSz)
+    )
   ]
 
 ------------------------------------------------------------------------
@@ -1469,6 +1486,67 @@ llvmVectorReduceUmin ::
         (BVType intSz)
 llvmVectorReduceUmin = llvmVectorReduce "umin" callVectorReduceUmin
 
+-- | Build an 'LLVMOverride' for a vector map intrinsic.
+llvmVectorMap ::
+  (1 <= intSz) =>
+  -- | The name of the operation to map (@umin@, @smin@, etc.).
+  String ->
+  -- | The semantics of the override.
+  (forall r args ret.
+    IsSymInterface sym =>
+    RegEntry sym (VectorType (BVType intSz)) ->
+    RegEntry sym (VectorType (BVType intSz)) ->
+    OverrideSim p sym ext r args ret (V.Vector (SymBV sym intSz))) ->
+  -- | The size of the vector type.
+  NatRepr vecSz ->
+  -- | The size of the integer type.
+  NatRepr intSz ->
+  LLVMOverride p sym ext
+    (EmptyCtx ::> VectorType (BVType intSz) ::> VectorType (BVType intSz))
+    (VectorType (BVType intSz))
+llvmVectorMap opName callMap vecSz intSz =
+  let nm = L.Symbol ("llvm." ++ opName ++
+                     ".v" ++ show (natValue vecSz) ++
+                     "i" ++ show (natValue intSz)) in
+    [llvmOvr| <#vecSz x #intSz> $nm( <#vecSz x #intSz>, <#vecSz x #intSz> ) |]
+    (\_memOps args -> Ctx.uncurryAssignment callMap args)
+
+llvmSmaxVector ::
+  (1 <= intSz) =>
+  NatRepr vecSz ->
+  NatRepr intSz ->
+  LLVMOverride p sym ext
+     (EmptyCtx ::> VectorType (BVType intSz) ::> VectorType (BVType intSz))
+     (VectorType (BVType intSz))
+llvmSmaxVector = llvmVectorMap "smax" callVectorMapSmax
+
+llvmSminVector ::
+  (1 <= intSz) =>
+  NatRepr vecSz ->
+  NatRepr intSz ->
+  LLVMOverride p sym ext
+     (EmptyCtx ::> VectorType (BVType intSz) ::> VectorType (BVType intSz))
+     (VectorType (BVType intSz))
+llvmSminVector = llvmVectorMap "smin" callVectorMapSmin
+
+llvmUmaxVector ::
+  (1 <= intSz) =>
+  NatRepr vecSz ->
+  NatRepr intSz ->
+  LLVMOverride p sym ext
+     (EmptyCtx ::> VectorType (BVType intSz) ::> VectorType (BVType intSz))
+     (VectorType (BVType intSz))
+llvmUmaxVector = llvmVectorMap "umax" callVectorMapUmax
+
+llvmUminVector ::
+  (1 <= intSz) =>
+  NatRepr vecSz ->
+  NatRepr intSz ->
+  LLVMOverride p sym ext
+     (EmptyCtx ::> VectorType (BVType intSz) ::> VectorType (BVType intSz))
+     (VectorType (BVType intSz))
+llvmUminVector = llvmVectorMap "umin" callVectorMapUmin
+
 ------------------------------------------------------------------------
 -- ** Implementations
 
@@ -2140,3 +2218,52 @@ callVectorReduceUmin intSz vec = do
   sym <- getSymInterface
   umax <- liftIO $ maxUnsignedBV sym intSz
   callVectorReduce (bvUmin sym) umax vec
+
+-- | The semantics of an LLVM vector map intrinsic.
+callVectorMap ::
+  -- | The operation to apply when mapping (e.g., @umin@, @smin@, etc.)
+  (RegValue sym tp -> RegValue sym tp -> IO (RegValue sym tp)) ->
+  -- | The first vector to map over.
+  RegEntry sym (VectorType tp) ->
+  -- | The second vector to map over.
+  RegEntry sym (VectorType tp) ->
+  -- | The result of mapping over the vectors.
+  OverrideSim p sym ext r args ret (V.Vector (RegValue sym tp))
+callVectorMap mapOp (regValue -> vec1) (regValue -> vec2) =
+  liftIO $ V.zipWithM mapOp vec1 vec2
+
+callVectorMapSmax ::
+  (IsSymInterface sym, 1 <= intSz) =>
+  RegEntry sym (VectorType (BVType intSz)) ->
+  RegEntry sym (VectorType (BVType intSz)) ->
+  OverrideSim p sym ext r args ret (V.Vector (SymBV sym intSz))
+callVectorMapSmax vec1 vec2 = do
+  sym <- getSymInterface
+  callVectorMap (bvSmax sym) vec1 vec2
+
+callVectorMapSmin ::
+  (IsSymInterface sym, 1 <= intSz) =>
+  RegEntry sym (VectorType (BVType intSz)) ->
+  RegEntry sym (VectorType (BVType intSz)) ->
+  OverrideSim p sym ext r args ret (V.Vector (SymBV sym intSz))
+callVectorMapSmin vec1 vec2 = do
+  sym <- getSymInterface
+  callVectorMap (bvSmin sym) vec1 vec2
+
+callVectorMapUmax ::
+  (IsSymInterface sym, 1 <= intSz) =>
+  RegEntry sym (VectorType (BVType intSz)) ->
+  RegEntry sym (VectorType (BVType intSz)) ->
+  OverrideSim p sym ext r args ret (V.Vector (SymBV sym intSz))
+callVectorMapUmax vec1 vec2 = do
+  sym <- getSymInterface
+  callVectorMap (bvUmax sym) vec1 vec2
+
+callVectorMapUmin ::
+  (IsSymInterface sym, 1 <= intSz) =>
+  RegEntry sym (VectorType (BVType intSz)) ->
+  RegEntry sym (VectorType (BVType intSz)) ->
+  OverrideSim p sym ext r args ret (V.Vector (SymBV sym intSz))
+callVectorMapUmin vec1 vec2 = do
+  sym <- getSymInterface
+  callVectorMap (bvUmin sym) vec1 vec2

